bson/src/main/org/bson/codecs/pojo/PojoIdGeneratingCodec.java

```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright 2008-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package org.bson.codecs.pojo;
+
+import org.bson.BsonDocument;
+import org.bson.BsonDocumentWriter;
+import org.bson.BsonObjectId;
+import org.bson.BsonReader;
+import org.bson.BsonValue;
+import org.bson.BsonWriter;
+import org.bson.codecs.Codec;
+import org.bson.codecs.DecoderContext;
+import org.bson.codecs.EncoderContext;
+import org.bson.codecs.configuration.CodecRegistry;
+import org.bson.types.ObjectId;
+
+import static org.bson.assertions.Assertions.isTrueArgument;
+import static org.bson.assertions.Assertions.notNull;
+
+
+class PojoIdGeneratingCodec<T, I> extends PojoCodec<T> {
```

Comment:
This seems to me less appropriate as a subclass than as a behavior modification of PojoCodecImpl.  Consider what we would do if there was some other behavior that we wanted to add that was orthogonal to id generation.  If we tried to make that a subclass as well, how would they be combined together?

I'm not sure the best way to fold this into PojoCodecImpl, but it seems straightforward enough that it could be done without making that class too bloated.  It's really just that one line in encode that would have to be modified.  The rest is additive and could live anywhere.

Reply:
This has been fixed and simplified.